### PR TITLE
Adds the RCT, A portable solution to problem removal.

### PR DESCRIPTION
**Pull request by ghost**
:cl: Floyd
add: the RCT, or Rapid Crate Transportation. A cargo traitor weapon for the cargo department. Do you hate your co-workers? do you wish they'd just go away? stuff them, or their corpse, into a locker or crate, close it and use the RCT on it. And BOOM! All your problems gone. To either space, or a location you set somewhere in the station in a now also welded closet, or crate.
/:cl:

Adds the RCT, a cargo department traitor weapon that allows you to teleport crates and closets, automatically welding them shut when you do.

activating / clicking on the RCT sets the location

alt clicking removes the location.

clicking on a closet or crate teleports the crate / locker to either your set location, or to a random location around the station in space if non was set.

costs 8 tc, only purchasable by cargo.

also yes i prevented abusing it to go to centcom or nullspace.

also changes the recycler obj/item into obj/, because it should be like that and else you cant recycle crates and lockers etc. Which was intended because else they wouldn't check if there's a human in the object it's getting the contents from.

### CODE REVIEW

File: code/game/objects/items/weapons/RCT.dm [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		user << "<span class='notice'>You start teleporting the closet</span>"
+		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+		if(do_after(user, 20, target = O))
+			var/obj/structure/closet/closet = O
+			user.visible_message("<span class='warning'>Sparks fly from [abbreviated_name] as the [closet] dissapears!</span>")
+			spark_system.start()
+			if(!setlocation)
+				for(var/obj/effect/landmark/C in landmarks_list)
+					if(C.name == "carpspawn")
+						closet.loc = C.loc
```
Comment (Time-Green) [inline code/game/objects/items/weapons/RCT.dm]: Use closet.forceMove(C) here

**Comment by Zostroll** [comment]
What is the delay on this? Like can you close a locker on someone then instantly send them to space in a welded locker?

File: code/game/objects/items/weapons/RCT.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			else
+				closet.loc = setlocation
```
Comment (Time-Green) [inline code/game/objects/items/weapons/RCT.dm]: And here.
It has some extra checks

Nice PR

**Comment by ghost** [comment]
1 secundo

**Comment by ghost** [comment]
do you really think I'd code something that would insta teleport any motherfucker in a locker instantly, that'd be so fucking OP 😆

**Comment by ghost** [comment]
the best you can do is use a locked locker but it's obvious as fuck if you do it. it is funny however.

**Review by Time-Green — COMMENTED** [review]
Really cool. I cant wait to hook this up to an emagged recycler

**Comment by ghost** [comment]
yeah i had to fix recyclers because of that, apparently they didn't recycle closets?

**Comment by Kmc2000** [comment]
This is really neat, add it

**Comment by Time-Green** [comment]
Should I worry about the recycler nullspacing people in lockers?

**Comment by ghost** [comment]
yes

**Comment by ghost** [comment]
and so should i

**Comment by ghost** [comment]
everything ok

**Comment by Jc1469** [comment]
So, this is an 8 tc item that has no cooldown, and takes a second to teleport a locker to a place, contents and all.  And by default is set to teleport to carp spawns.   

I..  :balance_scale: :police_car: 
Fuck that.

Stun a person, slip them, anything, put them in a locker, they're dead and gone before they can react.

**Comment by Time-Green** [comment]
It's 2 seconds

**Comment by ghost** [comment]
John can you read a page before commenting or do you just go on github to talk about your uneducated opinion because reading the description/comments/changed files is too difficult for you

**Comment by Jc1469** [comment]
>1 secundo

I think I covered it quite well, actually.

**Comment by Jc1469** [comment]
Oh wait, no, sorry. It also welds the locker.  I forgot that part.  It's still overpowered as fuck.

**Comment by ghost** [comment]
if you get slipped and put in a locker and someone has the time to close it and teleport you you are unrobust. Its also obvious as fuck if you do this and honestly by this logic the esword is op because you can slip and esword someone without them being able to react.

**Comment by Jc1469** [comment]
Yes, eswords are OP in that context.  Eswords dont teleport you into space though.  Or into a crusher.

**Comment by ghost** [comment]
Except you have to drag a closet around and be obvious as fuck about it if you want to legit kill someone with this. I hope you realize that someone that wants to robust someone fast and easy will just pick an esword instead of this because seriously slipping someone stuffing them in a locker and teleporting them is the most retarded way of going for a quick kill. Its also only available to the cargo department so its not even meta

**Comment by ghost** [comment]
it's entirely situational

it can land one kill every so often and is quite obvious in use,

it has a delay where simply slamming them in a locker or a brief slip won't cut it,

etc

File: code/game/objects/items/weapons/RCT.dm [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/obj/item/weapon/rct/attack_obj(obj/O, mob/living/user)
+	if(istype(O, /obj/structure/closet))
+		user << "<span class='notice'>You start teleporting the closet</span>"
+		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+		if(do_after(user, 20, target = O))
+			var/obj/structure/closet/closet = O
+			user.visible_message("<span class='warning'>Sparks fly from [abbreviated_name] as the [closet] dissapears!</span>")
+			spark_system.start()
+			if(!setlocation)
```
Comment (ShadowDeath6) [inline code/game/objects/items/weapons/RCT.dm]: I would provide feedback when this happens so they know a location hasn't been set yet.

**Comment by ShadowDeath6** [comment]
I think people are underestimating how long stuns are, considering this is a guaranteed kill on someone not wearing space protection. 
Aside from the obvious "yelling while I'm killing them" problem, you can stunprod someone, create a locker on top of them, stun them again, close it, then zap them into disposals, provided you're fast enough.

It's a pretty prep-heavy weapon, though, so I doubt it will be too much of a problem considering slipsword combo is still a thing.

**Comment by Cruix** [comment]
Eswords are very loud and everybody knows that their sound means "Someone is being violently murdered in the next room"

**Comment by ghost** [comment]
How about I make the delay 3 seconds? Remmeber that this item is cargo only, is situational (esword is always useful for example). This thing requires a locker and tbf if you can stun someone into a locker, close it and then also teleport them, without anyone seeing; you are either robust or the player is really fucktarded. Theres a reason most interesting antag items dont get used, its because theyre prematurely nerfed too hard.

**Comment by ghost** [comment]
5 seconds maybe

or make the sound loud and distinct

**Comment by ghost** [comment]
how long would the delay be allowed to be if it was loud and distinct

**Comment by ghost** [comment]
how about I scream into my mic thats pretty distinct

**Comment by Jc1469** [comment]
Tasers stun for 5 seconds.  Batons, prods too, stun for 7. Telescopics stun for 3.  Also did you know if you stun yourself with a telescopic baton as a clown it stuns you for 3*force, which equates to 30 seconds for a telebaton and 36 for a police baton.
And you can just chain the stuns until the end of time, too.

It's reasonably safe to say if you're trying to get someone with it, you've probably stunned them yourself, or are teleporting a body somewhere, so any noise that isn't just the machine placement noise, if you ask me.  pshoom_2 maybe. :ok_hand: 

As for timing.  I do want this item to be viable, believe it or not.  I just think 2 seconds after taking things like lag into account isn't a long enough time.  3 would probably be fine if the noise is distinct enough.

**Comment by ShadowDeath6** [comment]
this would function better as a "get this thing out of here discreetly" rather than a kill weapon, I think with a 3 second charge time it'll be good to use

File: code/game/objects/items/weapons/RCT.dm [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	src.spark_system = new /datum/effect_system/spark_spread
+	spark_system.set_up(5, 0, src)
+	spark_system.attach(src)
+
+/obj/item/weapon/rcd/Destroy()
+	qdel(spark_system)
+	spark_system = null
+
+/obj/item/weapon/rct/attack_self(mob/living/carbon/user)
+	if(istype(user/loc, /obj/structure/closet))
```
Comment (Cruix) [inline code/game/objects/items/weapons/RCT.dm]: this should be if(!isturf(user.loc)), so you cannot do crazy things like teleport people into mechs or disposal tubes.

Comment (Cruix) [inline code/game/objects/items/weapons/RCT.dm]: this should be if(!isturf(user.loc)), so you cannot do crazy things like teleport people into mechs or disposal tubes

File: code/game/objects/items/weapons/RCT.dm [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		user << "<span class='notice'>the [abbreviated_name] makes an angry noise.</span>"
+		playsound(src.loc, 'sound/machines/buzz-two.ogg', 50, 1)
+		return
+	setlocation = user.loc
+	playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+	user << "<span class='notice'>You set the teleportation location of the [abbreviated_name].</span>"
+	if(prob(20))
+		src.spark_system.start()
+
+/obj/item/weapon/rct/AltClick(mob/user)
```
Comment (Cruix) [inline code/game/objects/items/weapons/RCT.dm:31]: The description should tell you that alt-click does this

File: code/game/objects/items/weapons/RCT.dm [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+			spark_system.start()
+			if(!setlocation)
+				for(var/obj/effect/landmark/C in landmarks_list)
+					if(C.name == "carpspawn")
+						closet.forceMove(C)
+			else
+				closet.forceMove(setlocation)
+
+			spark_system.start()
+			closet.opened = FALSE
```
Comment (Cruix) [inline code/game/objects/items/weapons/RCT.dm]: This should be closet.close(), so the density updates, along with anything else specific to a particular type.

File: code/game/objects/items/weapons/RCT.dm [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+			if(istype(closet, /obj/structure/closet/crate))
+				var/obj/structure/closet/crate/crate = closet
+				if(crate.can_weld_shut)
```
Comment (Cruix) [inline code/game/objects/items/weapons/RCT.dm]: can_weld_shut is a closet var, so you can and should use whether it is a crate or a closet

File: code/game/objects/items/weapons/RCT.dm (hunk 4 above)
Comment (Cruix) [reply code/game/objects/items/weapons/RCT.dm]: Also I am amazed that this even compiles, user/loc makes no sense as a parameter

**Review by Cruix — CHANGES_REQUESTED** [review]

**Comment by Ratvar** [comment]
Hm. Allow it to print one holo-looking closet that last 30 seconds? You still can wrench chair nearby to have 50 metal sheets in five seconds, and people won't need to resist out of it when it goes off to designated supermatter chamber.

File: code/game/objects/items/weapons/RCT.dm (hunk 4 above)
Comment (ghost) [reply code/game/objects/items/weapons/RCT.dm]: the weirder thing was that it worked lmao

File: code/game/objects/items/weapons/RCT.dm (hunk 7 above)
Comment (Qustinnus) [reply code/game/objects/items/weapons/RCT.dm]: crates have no welding interaction I'm fairly sure.

**Comment by ghost** [comment]
added requested changes, 3 seconds timer now.

File: code/game/objects/items/weapons/RCT.dm [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	src.spark_system = new /datum/effect_system/spark_spread
+	spark_system.set_up(5, 0, src)
+	spark_system.attach(src)
+
+/obj/item/weapon/rcd/Destroy()
+	qdel(spark_system)
+	spark_system = null
+
+/obj/item/weapon/rct/attack_self(mob/living/carbon/user)
+	if(!isturf(user.loc)))
```
Comment (Jc1469) [inline code/game/objects/items/weapons/RCT.dm]: )))

Comment (ghost) [reply code/game/objects/items/weapons/RCT.dm]: lmao woops

**Comment by ghost** [comment]
ready for review I reckon.

**Comment by Cruix** [comment]
This will almost certainly need serious nerfs before it is actually merged.

**Comment by ghost** [comment]
alright, any ideas?

**Comment by Time-Green** [comment]
Wait couldn't I use this on the cargo shuttle, and then get to centcom?

**Comment by Cruix** [comment]
No, turf references do not move, even if the turf is moved by a shuttle.

**Comment by ghost** [comment]
ill ask again, any ideas to make it less OP?

**Comment by Cruix** [comment]
Make the delay longer
Make the delay longer for more or bigger items
Remove the ability to set the location it teleports to
Make it only work in cargo
Make it not weld the locker shut
Make it loud
Make it bulky
Make it not work on living things unless they are dead
Make it not work on living things at all
Make it have a long cooldown
Add commits with actual, informative commit messages

There are a hundred ways you could balance this, just do something so that people will not be slipped on water in the main hallway and be instantly, silently, uncounterably murdered.

**Comment by ghost** [comment]
image
credits to McDonald

**Comment by ghost** [comment]
feel free to finish this

**Pull request closed without merging**